src/analysis/souffle/examples/multimic/multimic-preferences.authlogic

```diff
@@ -0,0 +1,335 @@
+    isUserSpeech(UserX_userSpeech).
+
+  // The pipleine involving transcription and storage of the transcript is 
+  // allowed to run for a particular user if:
+  // 1) the user's preference is either permissive or that just transcripts can
+  // be stored.
+  // 2) the device owner's preference implies we should override the 
+  // preferences of others
+  // The rule "allow_transcription" captures these alternatives:
+  allowTranscription(UserX) :-
+    UserX hasPrivacyPreference("audio", "only_store_transcript").
+  allowTranscription(UserX) :- UserX hasPrivacyPreference("audio", "permissive").
+  allowTranscription(UserX) :- overridePreferences("audio"), isPrincipal(UserX).
+
+  "MicCompute" may("local_computation", UserX_userSpeech) :-
+    isUserSpeech(UserX_userSpeech), ownsTag(UserX, UserX_userSpeech),
+    allowTranscription(UserX).
+  
+  "MicCompute" canSay downgrades("asrStorageConn", UserX_userSpeech) :-
+    isUserSpeech(UserX_userSpeech), ownsTag(UserX, UserX_userSpeech),
+    allowTranscription(UserX).
+
+  "MicCompute" canSay hasTag("asrStorageConn", UserX_speechTranscript) :-
+    isUserTranscript(UserX_speechTranscript), ownsTag(UserX, UserX_speechTranscript),
+    allowTranscription(UserX).
+  
+  "SpeechToTextStorageSink" may("store", UserX_speechTranscript) :-
+    isUserTranscript(UserX_speechTranscript), ownsTag(UserX, UserX_speechTranscript),
+    allowTranscription(UserX).
+
+  // Should be generated behind the scenes:
+  isPrincipal("UserA").
+  isPrincipal("UserB").
+  isPrincipal("UserC").
+  isPrivacyPreference("permissive").
+  isPrivacyPreference("only_store_transcript").
+  isPrivacyPreference("only_notify").
+  isPrivacyResolution("stay_on").
+  isPrivacyResolution("meet_requests").
+  isUserSpeech("UserA_userSpeech").
+  isUserSpeech("UserB_userSpeech").
+  isUserSpeech("UserC_userSpeech").
+  ownsTag("UserA", "UserA_userSpeech").
+  ownsTag("UserB", "UserB_userSpeech").
+  ownsTag("UserC", "UserC_userSpeech").
+  isUserTranscript("UserA_speechTranscript").
+  isUserTranscript("UserB_speechTranscript").
+  isUserTranscript("UserC_speechTranscript").
+  ownsTag("UserA", "UserA_speechTranscript").
+  ownsTag("UserB", "UserB_speechTranscript").
+  ownsTag("UserC", "UserC_speechTranscript").
+}
+
+//-----------------------------------------------------------------------------
+// "Will" Expressions (that go near particles)
+//-----------------------------------------------------------------------------
+// These statements are also kept near the particles, but in order to actually 
+// run this they are repeated here. These are used to generate the checks 
+// against the permissions granted with "may"
+"Notifier" says {
+    will("deliver_privacy_notifications", "audioIn").
+}
+
+"MicCompute" says {
+    will("local_computation", "audioIn"). 
+    // Assumes we can figure out tags with namespaces like this:
+    downgrades("asrStorageConn", UserX_userSpeech) :- 
```

Comment:
@bgogul here was another way to express downgrading speech by all the users -- make a relation for the type of data (userSpeech) and add all the user speech tags to this relation. Populating the relation wasn't done yet, though, so the ownership thing is fine.

Reply:
@markww not sure if you noticed this actually.

Reply:
@markww not sure if you noticed this in the code -- you can downgrade all the data of a particular "type" by packing it into a relation for that type. I used this as a workaround for not having namespaces, which I used magically in the other PR before this one.